File: src/contracts/entities/cluster.ts
import type { ClusterStatus } from '../enums/cluster-status';

/**
 * Cluster — Unidade raiz do sistema.
 * Representa uma familia com identidade criptografica derivada de seed phrase BIP-39.
 * cluster_id = SHA-256(public_key) — imutavel apos criacao (RN-C1).
 * Limites: max 10 membros (RN-C3), max 50 nos (RN-C4).
 */
export interface Cluster {
  id: string;
  clusterId: string;
  name: string;
  publicKey: Uint8Array;
  encryptedPrivateKey: Uint8Array;
  status: ClusterStatus;
  createdAt: Date;
  updatedAt: Date;
}

File: crates/core-sdk/src/domain/cluster.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Grupo familiar que compartilha armazenamento distribuido.
/// Entidade raiz que agrupa membros, nos e arquivos.
/// Possui identidade criptografica derivada de um par de chaves.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Cluster {
    pub id: Uuid,
    /// Hash da chave publica — identidade criptografica imutavel
    pub cluster_id: String,
    pub name: String,
    /// Chave publica para verificacao de assinaturas
    pub public_key: Vec<u8>,
    /// Chave privada criptografada com master key
    pub encrypted_private_key: Vec<u8>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

// -- Eventos de dominio --

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ClusterEvent {
    /// Dispara geracao de seed phrase e criacao do vault do membro admin
    Created { cluster_id: String },
    /// Dispara reconexao de nos e rebuild do indice de metadados
    Recovered { cluster_id: String },
}
